test/unit-test/test_simulator.c

```diff
@@ -0,0 +1,155 @@
+// Copyright 2023-2024 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "hww.h"
+#include "memory/bitbox02_smarteeprom.h"
+#include "usb/usb_packet.c"
+#include "usb/usb_processing.c"
+#include "usb/usb_processing.h"
+#include "workflow/idle_workflow.h"
+#include <fcntl.h>
+#include <memory/memory.h>
+#include <mock_memory.h>
+#include <queue.h>
+#include <random.h>
+#include <rust/rust.h>
+#include <sd.h>
+#include <stdio.h>
+#include <unistd.h>
+
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#define BUFFER_SIZE 1024
+#define SOCKET
+
+static uint8_t _out_report[USB_HID_REPORT_OUT_SIZE];
+char out_buffer[2 * USB_HID_REPORT_OUT_SIZE];
+int data_len;
+int sockfd;
+
+int get_usb_message_socket(char* buffer, uint8_t* input)
+{
+    int num_read = read(sockfd, buffer, USB_HID_REPORT_OUT_SIZE);
+    for (int i = 0; i < USB_HID_REPORT_OUT_SIZE; i++) {
+        input[i] = (uint8_t)buffer[i];
+    }
+    return num_read;
+}
+
+void send_usb_message_socket(void)
+{
+    struct queue* q = queue_hww_queue();
+    const uint8_t* data = queue_pull(q);
+    if (data != NULL) {
+        data_len = 256 * (int)data[5] + (int)data[6];
+        for (int i = 0; i < USB_HID_REPORT_OUT_SIZE; i++) {
+            out_buffer[i] = (char)data[i];
+        }
+        if (!write(sockfd, out_buffer, USB_HID_REPORT_OUT_SIZE)) {
+            perror("ERROR, could not write to socket");
+            exit(1);
+        }
+    }
+}
+
+void simulate_firmware_execution(uint8_t* input)
+{
+    memcpy(_out_report, input, sizeof(_out_report));
+    usb_packet_process((const USB_FRAME*)_out_report);
+    rust_workflow_spin();
+    rust_async_usb_spin();
+    usb_processing_process(usb_processing_hww());
+}
+
+int main(void)
+{
+    char buffer[BUFFER_SIZE];
+    uint8_t input[BUFFER_SIZE];
+    int (*get_message)(char*, uint8_t*);
+    void (*send_message)(void);
+    bool memory_success, sd_success;
+    int temp_len;
```

Comment:
no need to declare the vars at the top, code is imo nicer when the vars are just declared on first use, since we don't care much about stack use.  `bool memory_success = memory_setup(...)` etc.